File: lib-database/src/main/java/com/ustadmobile/core/db/dao/UMCalendarDao.java
package com.ustadmobile.core.db.dao;

import com.ustadmobile.core.db.UmLiveData;
import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.lib.database.annotation.UmDao;
import com.ustadmobile.lib.database.annotation.UmInsert;
import com.ustadmobile.lib.database.annotation.UmQuery;
import com.ustadmobile.lib.database.annotation.UmRepository;
import com.ustadmobile.lib.database.annotation.UmUpdate;
import com.ustadmobile.lib.db.entities.UMCalendar;
import com.ustadmobile.lib.db.entities.UMCalendarWithNumEntries;
import com.ustadmobile.lib.db.sync.dao.SyncableDao;

import java.util.List;

@UmDao(updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN,
insertPermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@UmRepository
public abstract class UMCalendarDao implements SyncableDao<UMCalendar, UMCalendarDao> {

    @UmInsert
    public abstract long insert(UMCalendar entity);

    @UmUpdate
    public abstract void update(UMCalendar entity);

    @UmInsert
    public abstract void insertAsync(UMCalendar entity, UmCallback<Long> result);

    @UmQuery("SELECT * FROM UMCalendar WHERE umCalendarActive = 1 AND " +
            " umCalendarCategory = " + UMCalendar.CATEGORY_HOLIDAY )
    public abstract UmProvider<UMCalendar> findAllHolidays();

    @UmQuery("SELECT * , 0 AS numEntries FROM UMCalendar WHERE umCalendarActive = 1 AND " +
            " umCalendarCategory = " + UMCalendar.CATEGORY_HOLIDAY )
    public abstract UmProvider<UMCalendarWithNumEntries> findAllHolidaysWithEntriesCount();

    @UmQuery("SELECT * FROM UMCalendar")
    public abstract UmProvider<UMCalendar> findAllUMCalendars();

    @UmQuery("SELECT * FROM UMCalendar")
    public abstract UmLiveData<List<UMCalendar>> findAllUMCalendarsAsLiveDataList();

    @UmQuery("SELECT * FROM UMCalendar WHERE umCalendarCategory = " + UMCalendar.CATEGORY_HOLIDAY )
    public abstract UmLiveData<List<UMCalendar>> findAllHolidaysLiveData();

    @UmQuery("SELECT * FROM UMCalendar WHERE umCalendarUid = :uid")
    public abstract UmLiveData<UMCalendar> findByUidLive(long uid);

    @UmUpdate
    public abstract void updateAsync(UMCalendar entity, UmCallback<Integer> result);

    @UmQuery("SELECT * FROM UMCalendar WHERE umCalendarName = :name")
    public abstract UMCalendar findByName(String name);

    @UmQuery("SELECT * FROM UMCalendar WHERE umCalendarUid = :uid")
    public abstract UMCalendar findByUid(long uid);

    @UmQuery("SELECT * FROM UMCalendar WHERE umCalendarUid = :uid")
    public abstract void findByUidAsync(long uid, UmCallback<UMCalendar> resultObject);

}

File: lib-database/src/commonMain/kotlin/com/ustadmobile/core/db/dao/UMCalendarDao.kt
package com.ustadmobile.core.db.dao

import androidx.paging.DataSource
import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import com.ustadmobile.core.impl.UmCallback
import com.ustadmobile.door.DoorLiveData
import com.ustadmobile.lib.database.annotation.UmDao
import com.ustadmobile.lib.database.annotation.UmRepository
import com.ustadmobile.lib.db.entities.UMCalendar
import com.ustadmobile.lib.db.entities.UMCalendarWithNumEntries

@UmDao(updatePermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN, 
        insertPermissionCondition = RoleDao.SELECT_ACCOUNT_IS_ADMIN)
@UmRepository
@Dao
abstract class UMCalendarDao : BaseDao<UMCalendar> {

    @Insert
    abstract override fun insert(entity: UMCalendar): Long

    @Update
    abstract override fun update(entity: UMCalendar)

    @Query("SELECT * FROM UMCalendar WHERE CAST(umCalendarActive AS INTEGER) = 1 AND " +
            " umCalendarCategory = " + UMCalendar.CATEGORY_HOLIDAY)
    abstract fun findAllHolidays(): DataSource.Factory<Int, UMCalendar>

    @Query("SELECT UMCalendar.* ," +
            " (SELECT COUNT(*) FROM DateRange " +
            "   WHERE dateRangeUMCalendarUid = UMCalendar.umCalendarUid " +
            "   AND dateRange.dateRangeActive = 1) AS numEntries " +
            " FROM UMCalendar WHERE CAST(umCalendarActive AS INTEGER) = 1 AND " +
            " umCalendarCategory = " + UMCalendar.CATEGORY_HOLIDAY)
    abstract fun findAllHolidaysWithEntriesCount(): DataSource.Factory<Int, UMCalendarWithNumEntries>

    @Query("SELECT * FROM UMCalendar WHERE CAST(umCalendarActive AS INTEGER) = 1")
    abstract fun findAllUMCalendars(): DataSource.Factory<Int, UMCalendar>

    @Query("SELECT * FROM UMCalendar WHERE CAST(umCalendarActive AS INTEGER) = 1")
    abstract fun findAllUMCalendarsAsLiveDataList(): DoorLiveData<List<UMCalendar>>

    @Query("SELECT * FROM UMCalendar WHERE CAST(umCalendarActive AS INTEGER) = 1 AND umCalendarCategory = "
            + UMCalendar.CATEGORY_HOLIDAY)
    abstract fun findAllHolidaysLiveData(): DoorLiveData<List<UMCalendar>>

    @Query("SELECT * FROM UMCalendar WHERE umCalendarUid = :uid AND CAST(umCalendarActive AS INTEGER) = 1")
    abstract fun findByUidLive(uid: Long): DoorLiveData<UMCalendar?>

    @Update
    abstract suspend fun updateAsync(entity: UMCalendar):Int

    @Query("SELECT * FROM UMCalendar WHERE umCalendarName = :name")
    abstract fun findByName(name: String): UMCalendar?

    @Query("SELECT * FROM UMCalendar WHERE umCalendarUid = :uid")
    abstract fun findByUid(uid: Long): UMCalendar?

    @Query("SELECT * FROM UMCalendar WHERE umCalendarUid = :uid")
    abstract suspend fun findByUidAsync(uid: Long): UMCalendar?

    @Query("UPDATE UMCalendar SET umCalendarActive = 0 WHERE umCalendarUid = :uid")
    abstract suspend fun inactivateCalendarAsync(uid: Long): Int

}
